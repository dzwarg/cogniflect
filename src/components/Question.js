import React from 'react';
import StartOver from '../containers/StartOver';
import {Button, Col, Container, Form, Jumbotron, ProgressBar, Row} from 'react-bootstrap';
import {LinkContainer} from 'react-router-bootstrap';

const Question = ({appState, changeHandler, history, match}) => {
  const questionId = parseInt(match.params.questionId, 10) - 1;
  const questionData = appState.get('questions').get(questionId);
  const questionDataJS = questionData.toJS();
  const questionSize = appState.get('questions').size;
  const cursor = 100 * (questionId / questionSize);
  const memberAnswers = appState.getIn(['questions', questionId, 'teamAnswerCount']);
  const totalMembers = appState.getIn(['team', 'members']);
  const assessmentType = appState.get('assessmentType');
  const answerKey = (assessmentType === 'individual') ? 'myAnswer' : 'ourAnswer';
  
  const collabText = (assessmentType === 'team') ?
    (<p>{memberAnswers} of {totalMembers} team members have answered this question.</p>) :
    null
  const nextLocation = (questionId + 1) === questionSize ? '/summary' : `/question/${questionDataJS.next + 1}`;
  const startOverChildren = (assessmentType === 'team') ? 
        (<LinkContainer to={nextLocation}>
          {
            memberAnswers === totalMembers ?
            <Button bsStyle="success">Next</Button> :
            <Button bsStyle="success" disabled>Next</Button>
          }
        </LinkContainer>) :
        null;
  return (
    <Container>
      <Row>
        <Col>
          <Jumbotron>
            <h1 className="capitalized">{assessmentType} assessment</h1>
            {collabText}
          </Jumbotron>
        </Col>
      </Row>
      <Row>
        <Col xs={12}>
          <h2>Question #{match.params.questionId}</h2>
        </Col>
      </Row>
      <Row>
        <Col xs={12}>
          <Form>
            <p>
              {questionDataJS.text}
            </p>
            <ul>
              {questionDataJS.guesses.map((guess) => {
                const checkProps = {
                  onChange: changeHandler(questionData, questionId, assessmentType, answerKey, guess.id, history),
                  checked: (guess.id === questionDataJS[answerKey])
                }
    
                return (
                 <li key={guess.id}><Form.Check type="checkbox" {...checkProps}>{guess.text}</Form.Check></li>
                );
              })}
            </ul>
          </Form>
        </Col>
      </Row>
      <Row>
        <Col xs={12}>
          <ProgressBar now={cursor} label={`${cursor}%`} srOnly active/>
        </Col>
      </Row>
      <StartOver>
        {startOverChildren}
      </StartOver>
    </Container>
  );
};

export default Question;