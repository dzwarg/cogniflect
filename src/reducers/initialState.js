import {fromJS} from 'immutable';

const initialState = fromJS({
  teamCode: '',
  userTeamCode: '',
  assessmentType: 'individual',
  questionCursor: 0,
  questions: [
    {
      id: 0,
      text: "In a lake, there is a patch of lily pads. Every day, the patch doubles in size. If it takes 48 days for the patch to cover the entire lake, how long would it take for the patch to cover half of the lake?",
      truth: 1,
      myAnswer: null,
      ourAnswer: null,
      teamAnswerCount: 0,
      guesses: [
        { id: 0, text: '24' },
        { id: 1, text: '47' },
        { id: 2, text: '96' },
        { id: 3, text: '49' }
      ],
      next: 1
    },
    {
      id: 1,
      text: "A bat and a ball cost $1.10 in total. The bat costs $1.00 more than the ball. How much does the ball cost?",
      truth: 0,
      myAnswer: null,
      ourAnswer: null,
      teamAnswerCount: 0,
      guesses: [
        { id: 0, text: '$0.05' },
        { id: 1, text: '$0.06' },
        { id: 2, text: '$0.10' },
        { id: 3, text: '$0.50' }
      ],
      next: 2
    },
    {
      id: 2,
      text: "If it takes 5 machines 5 minutes to make 5 widgets, how long would it take 100 machines to make 100 widgets?",
      truth: 0,
      myAnswer: null,
      ourAnswer: null,
      teamAnswerCount: 0,
      guesses: [
        { id: 0, text: '5 minutes' },
        { id: 1, text: '100 minutes' },
        { id: 2, text: '500 minutes' },
        { id: 3, text: '10000 minutes' }
      ],
      next: 3
    },
    {
      id: 3,
      text: "If you’re running a race and you pass the person in second place, what place are you in?",
      truth: 1,
      myAnswer: null,
      ourAnswer: null,
      teamAnswerCount: 0,
      guesses: [
        { id: 0, text: '1st place' },
        { id: 1, text: '2nd place' },
        { id: 2, text: '3rd place' },
        { id: 3, text: 'Last place' }
      ],
      next: 4
    },
    {
      id: 4,
      text: "A farmer had 15 sheep and all but 8 died. How many are left?",
      truth: 2,
      myAnswer: null,
      ourAnswer: null,
      teamAnswerCount: 0,
      guesses: [
        { id: 0, text: '6' },
        { id: 1, text: '7' },
        { id: 2, text: '8' },
        { id: 3, text: '9' }
      ],
      next: 5
    },
    {
      id: 5,
      text: "Emily’s father has three daughters. The first two are named April and May. What is the third daughter’s name?",
      truth: 3,
      myAnswer: null,
      ourAnswer: null,
      teamAnswerCount: 0,
      guesses: [
        { id: 0, text: 'June' },
        { id: 1, text: 'August' },
        { id: 2, text: 'Summer' },
        { id: 3, text: 'Emily' }
      ],
      next: 6
    },
    {
      id: 6,
      text: "How many cubic feet of dirt are there in a hole that is 3’ deep x 3’ wide x 3’ long?",
      truth: 0,
      myAnswer: null,
      ourAnswer: null,
      teamAnswerCount: 0,
      guesses: [
        { id: 0, text: '0 ft^3' },
        { id: 1, text: '3 ft^3' },
        { id: 2, text: '9 ft^3' },
        { id: 3, text: '27 ft^3' }
      ],
      next: null
    }
  ],
  team: {
    members: 0,
    synced: 0
  }
});

export default initialState;