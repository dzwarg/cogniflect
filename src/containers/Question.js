import { connect } from 'react-redux';
import actions from '../actions';
import Question from '../components/Question';
 
const mapStateToProps = (state) => ({
  appState: state.appState
});

const mapDispatchToProps = (dispatch) => ({
  changeHandler: (questionData, questionId, assessmentType, answerKey, guessId, history) => (evt) => {
    evt.preventDefault();
    
    dispatch(actions.set(['questions', questionId], questionData.set(answerKey, guessId)));

    if (assessmentType === 'individual') {
      const nextQuestion = questionData.get('next');
      dispatch(actions.set(['questionCursor'], nextQuestion))
      if (nextQuestion !== null) {
        history.push(`/question/${nextQuestion + 1}`);
      } else {
        history.push('/synchronize');
      }
    }
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(Question)
export {mapStateToProps, mapDispatchToProps}