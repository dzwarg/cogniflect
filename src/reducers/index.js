import { combineReducers } from 'redux'
import initialState from './initialState'; 

const appState = (state = initialState, action) => {
  switch (action.type) {
    case 'SET':
      return state.setIn(action.payload.path, action.payload.value);
    default:
      return state
  }
}

export default combineReducers({
  appState
});