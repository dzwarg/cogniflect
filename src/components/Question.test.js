import React from 'react';
import Question from './Question';
import {Jumbotron} from 'react-bootstrap';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import reducers from '../reducers';
import {fromJS} from 'immutable';

const initialState = {
  appState: fromJS({
    questions: [{
      id: 0,
      text: "In a lake, there is a patch of lily pads. Every day, the patch doubles in size. If it takes 48 days for the patch to cover the entire lake, how long would it take for the patch to cover half of the lake?",
      truth: 1,
      myAnswer: null,
      ourAnswer: null,
      teamAnswerCount: 0,
      guesses: [
        { id: 0, text: '24' },
        { id: 1, text: '47' },
        { id: 2, text: '96' },
        { id: 3, text: '49' }
      ],
      next: 1
    },{
      id: 1,
      text: "A bat and a ball cost $1.10 in total. The bat costs $1.00 more than the ball. How much does the ball cost?",
      truth: 0,
      myAnswer: null,
      ourAnswer: null,
      teamAnswerCount: 0,
      guesses: [
        { id: 0, text: '$0.05' },
        { id: 1, text: '$0.06' },
        { id: 2, text: '$0.10' },
        { id: 3, text: '$0.50' }
      ],
      next: null
    }],
    team: {
      members: 0,
      synced: 0
    },
    assessmentType: 'individual'
  })
};

describe('components/Question', () => {
  it('matches snapshot for individual', () => {
    const component = Question({
      ...initialState,
      changeHandler: ()=>{},
      match: {params:{questionId: 1}}
    })
    expect(component).toMatchSnapshot();
  });

  it('matches snapshot for team', () => {
    const teamState = {
      appState: initialState.appState.set('assessmentType', 'team')
    };
    const component = Question({
      ...teamState,
      changeHandler:()=>{},
      match:{params:{questionId: 1}}
    })
    expect(component).toMatchSnapshot();
  });

  it('matches snapshot for team last question', () => {
    const teamState = {
      appState: initialState.appState
        .set('assessmentType', 'team')
    };
    const component = Question({
      ...teamState,
      changeHandler: ()=>{},
      match: {params:{questionId: 2}}
    })
    expect(component).toMatchSnapshot();
  });

  it('matches snapshot for team all answered', () => {
    const teamState = {
      appState: initialState.appState
        .set('assessmentType', 'team')
        .setIn(['team', 'members'], 1)
    };
    const component = Question({
      ...teamState,
      changeHandler: ()=>{},
      match: {params:{questionId: 1}}
    })
    expect(component).toMatchSnapshot();
  });
});