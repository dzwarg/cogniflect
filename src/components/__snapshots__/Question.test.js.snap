// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dumb component renders for individual 1`] = `
<ForwardRef(Bootstrap(Container))>
  <ForwardRef(Bootstrap(Row))>
    <ForwardRef(Bootstrap(Col))>
      <ForwardRef(Bootstrap(Jumbotron))>
        <h1
          className="capitalized"
        >
          individual
           assessment
        </h1>
      </ForwardRef(Bootstrap(Jumbotron))>
    </ForwardRef(Bootstrap(Col))>
  </ForwardRef(Bootstrap(Row))>
  <ForwardRef(Bootstrap(Row))>
    <ForwardRef(Bootstrap(Col))
      xs={12}
    >
      <h2>
        Question #
        1
      </h2>
    </ForwardRef(Bootstrap(Col))>
  </ForwardRef(Bootstrap(Row))>
  <ForwardRef(Bootstrap(Row))>
    <ForwardRef(Bootstrap(Col))
      xs={12}
    >
      <ForwardRef(Bootstrap(Form))>
        <p>
          In a lake, there is a patch of lily pads. Every day, the patch doubles in size. If it takes 48 days for the patch to cover the entire lake, how long would it take for the patch to cover half of the lake?
        </p>
        <ul>
          <li>
            <ForwardRef(Bootstrap(FormCheck))
              checked={false}
              type="checkbox"
            >
              24
            </ForwardRef(Bootstrap(FormCheck))>
          </li>
          <li>
            <ForwardRef(Bootstrap(FormCheck))
              checked={false}
              type="checkbox"
            >
              47
            </ForwardRef(Bootstrap(FormCheck))>
          </li>
          <li>
            <ForwardRef(Bootstrap(FormCheck))
              checked={false}
              type="checkbox"
            >
              96
            </ForwardRef(Bootstrap(FormCheck))>
          </li>
          <li>
            <ForwardRef(Bootstrap(FormCheck))
              checked={false}
              type="checkbox"
            >
              49
            </ForwardRef(Bootstrap(FormCheck))>
          </li>
        </ul>
      </ForwardRef(Bootstrap(Form))>
    </ForwardRef(Bootstrap(Col))>
  </ForwardRef(Bootstrap(Row))>
  <ForwardRef(Bootstrap(Row))>
    <ForwardRef(Bootstrap(Col))
      xs={12}
    >
      <ForwardRef(Bootstrap(ProgressBar))
        active={true}
        label="0%"
        now={0}
        srOnly={true}
      />
    </ForwardRef(Bootstrap(Col))>
  </ForwardRef(Bootstrap(Row))>
  <Connect(StartOver) />
</ForwardRef(Bootstrap(Container))>
`;

exports[`dumb component renders for team 1`] = `
<ForwardRef(Bootstrap(Container))>
  <ForwardRef(Bootstrap(Row))>
    <ForwardRef(Bootstrap(Col))>
      <ForwardRef(Bootstrap(Jumbotron))>
        <h1
          className="capitalized"
        >
          team
           assessment
        </h1>
        <p>
          0
           of 
          0
           team members have answered this question.
        </p>
      </ForwardRef(Bootstrap(Jumbotron))>
    </ForwardRef(Bootstrap(Col))>
  </ForwardRef(Bootstrap(Row))>
  <ForwardRef(Bootstrap(Row))>
    <ForwardRef(Bootstrap(Col))
      xs={12}
    >
      <h2>
        Question #
        1
      </h2>
    </ForwardRef(Bootstrap(Col))>
  </ForwardRef(Bootstrap(Row))>
  <ForwardRef(Bootstrap(Row))>
    <ForwardRef(Bootstrap(Col))
      xs={12}
    >
      <ForwardRef(Bootstrap(Form))>
        <p>
          In a lake, there is a patch of lily pads. Every day, the patch doubles in size. If it takes 48 days for the patch to cover the entire lake, how long would it take for the patch to cover half of the lake?
        </p>
        <ul>
          <li>
            <ForwardRef(Bootstrap(FormCheck))
              checked={false}
              type="checkbox"
            >
              24
            </ForwardRef(Bootstrap(FormCheck))>
          </li>
          <li>
            <ForwardRef(Bootstrap(FormCheck))
              checked={false}
              type="checkbox"
            >
              47
            </ForwardRef(Bootstrap(FormCheck))>
          </li>
          <li>
            <ForwardRef(Bootstrap(FormCheck))
              checked={false}
              type="checkbox"
            >
              96
            </ForwardRef(Bootstrap(FormCheck))>
          </li>
          <li>
            <ForwardRef(Bootstrap(FormCheck))
              checked={false}
              type="checkbox"
            >
              49
            </ForwardRef(Bootstrap(FormCheck))>
          </li>
        </ul>
      </ForwardRef(Bootstrap(Form))>
    </ForwardRef(Bootstrap(Col))>
  </ForwardRef(Bootstrap(Row))>
  <ForwardRef(Bootstrap(Row))>
    <ForwardRef(Bootstrap(Col))
      xs={12}
    >
      <ForwardRef(Bootstrap(ProgressBar))
        active={true}
        label="0%"
        now={0}
        srOnly={true}
      />
    </ForwardRef(Bootstrap(Col))>
  </ForwardRef(Bootstrap(Row))>
  <Connect(StartOver)>
    <LinkContainer
      activeClassName="active"
      exact={false}
      replace={false}
      strict={false}
      to="/question/2"
    >
      <ForwardRef(Bootstrap(Button))
        bsStyle="success"
      >
        Next
      </ForwardRef(Bootstrap(Button))>
    </LinkContainer>
  </Connect(StartOver)>
</ForwardRef(Bootstrap(Container))>
`;

exports[`dumb component renders team all answered 1`] = `
<ForwardRef(Bootstrap(Container))>
  <ForwardRef(Bootstrap(Row))>
    <ForwardRef(Bootstrap(Col))>
      <ForwardRef(Bootstrap(Jumbotron))>
        <h1
          className="capitalized"
        >
          team
           assessment
        </h1>
        <p>
          0
           of 
          1
           team members have answered this question.
        </p>
      </ForwardRef(Bootstrap(Jumbotron))>
    </ForwardRef(Bootstrap(Col))>
  </ForwardRef(Bootstrap(Row))>
  <ForwardRef(Bootstrap(Row))>
    <ForwardRef(Bootstrap(Col))
      xs={12}
    >
      <h2>
        Question #
        1
      </h2>
    </ForwardRef(Bootstrap(Col))>
  </ForwardRef(Bootstrap(Row))>
  <ForwardRef(Bootstrap(Row))>
    <ForwardRef(Bootstrap(Col))
      xs={12}
    >
      <ForwardRef(Bootstrap(Form))>
        <p>
          In a lake, there is a patch of lily pads. Every day, the patch doubles in size. If it takes 48 days for the patch to cover the entire lake, how long would it take for the patch to cover half of the lake?
        </p>
        <ul>
          <li>
            <ForwardRef(Bootstrap(FormCheck))
              checked={false}
              type="checkbox"
            >
              24
            </ForwardRef(Bootstrap(FormCheck))>
          </li>
          <li>
            <ForwardRef(Bootstrap(FormCheck))
              checked={false}
              type="checkbox"
            >
              47
            </ForwardRef(Bootstrap(FormCheck))>
          </li>
          <li>
            <ForwardRef(Bootstrap(FormCheck))
              checked={false}
              type="checkbox"
            >
              96
            </ForwardRef(Bootstrap(FormCheck))>
          </li>
          <li>
            <ForwardRef(Bootstrap(FormCheck))
              checked={false}
              type="checkbox"
            >
              49
            </ForwardRef(Bootstrap(FormCheck))>
          </li>
        </ul>
      </ForwardRef(Bootstrap(Form))>
    </ForwardRef(Bootstrap(Col))>
  </ForwardRef(Bootstrap(Row))>
  <ForwardRef(Bootstrap(Row))>
    <ForwardRef(Bootstrap(Col))
      xs={12}
    >
      <ForwardRef(Bootstrap(ProgressBar))
        active={true}
        label="0%"
        now={0}
        srOnly={true}
      />
    </ForwardRef(Bootstrap(Col))>
  </ForwardRef(Bootstrap(Row))>
  <Connect(StartOver)>
    <LinkContainer
      activeClassName="active"
      exact={false}
      replace={false}
      strict={false}
      to="/question/2"
    >
      <ForwardRef(Bootstrap(Button))
        bsStyle="success"
        disabled={true}
      >
        Next
      </ForwardRef(Bootstrap(Button))>
    </LinkContainer>
  </Connect(StartOver)>
</ForwardRef(Bootstrap(Container))>
`;

exports[`dumb component renders team last question 1`] = `
<ForwardRef(Bootstrap(Container))>
  <ForwardRef(Bootstrap(Row))>
    <ForwardRef(Bootstrap(Col))>
      <ForwardRef(Bootstrap(Jumbotron))>
        <h1
          className="capitalized"
        >
          team
           assessment
        </h1>
        <p>
          0
           of 
          0
           team members have answered this question.
        </p>
      </ForwardRef(Bootstrap(Jumbotron))>
    </ForwardRef(Bootstrap(Col))>
  </ForwardRef(Bootstrap(Row))>
  <ForwardRef(Bootstrap(Row))>
    <ForwardRef(Bootstrap(Col))
      xs={12}
    >
      <h2>
        Question #
        2
      </h2>
    </ForwardRef(Bootstrap(Col))>
  </ForwardRef(Bootstrap(Row))>
  <ForwardRef(Bootstrap(Row))>
    <ForwardRef(Bootstrap(Col))
      xs={12}
    >
      <ForwardRef(Bootstrap(Form))>
        <p>
          A bat and a ball cost $1.10 in total. The bat costs $1.00 more than the ball. How much does the ball cost?
        </p>
        <ul>
          <li>
            <ForwardRef(Bootstrap(FormCheck))
              checked={false}
              type="checkbox"
            >
              $0.05
            </ForwardRef(Bootstrap(FormCheck))>
          </li>
          <li>
            <ForwardRef(Bootstrap(FormCheck))
              checked={false}
              type="checkbox"
            >
              $0.06
            </ForwardRef(Bootstrap(FormCheck))>
          </li>
          <li>
            <ForwardRef(Bootstrap(FormCheck))
              checked={false}
              type="checkbox"
            >
              $0.10
            </ForwardRef(Bootstrap(FormCheck))>
          </li>
          <li>
            <ForwardRef(Bootstrap(FormCheck))
              checked={false}
              type="checkbox"
            >
              $0.50
            </ForwardRef(Bootstrap(FormCheck))>
          </li>
        </ul>
      </ForwardRef(Bootstrap(Form))>
    </ForwardRef(Bootstrap(Col))>
  </ForwardRef(Bootstrap(Row))>
  <ForwardRef(Bootstrap(Row))>
    <ForwardRef(Bootstrap(Col))
      xs={12}
    >
      <ForwardRef(Bootstrap(ProgressBar))
        active={true}
        label="50%"
        now={50}
        srOnly={true}
      />
    </ForwardRef(Bootstrap(Col))>
  </ForwardRef(Bootstrap(Row))>
  <Connect(StartOver)>
    <LinkContainer
      activeClassName="active"
      exact={false}
      replace={false}
      strict={false}
      to="/summary"
    >
      <ForwardRef(Bootstrap(Button))
        bsStyle="success"
      >
        Next
      </ForwardRef(Bootstrap(Button))>
    </LinkContainer>
  </Connect(StartOver)>
</ForwardRef(Bootstrap(Container))>
`;
